pgscatalog.match/src/pgscatalog/match/lib/_match/plink.py

```diff
@@ -143,44 +145,55 @@ def _deduplicate_variants(df: pl.LazyFrame) -> list[pl.LazyFrame]:
     Returns:
         A list of dataframes, with unique ID - matched effect allele combinations
     """
-    if df.limit(1).collect().is_empty():
-        logger.info("Empty input: skipping deduplication")
+    if df.select(pl.len()).collect().item(0, 0) == 0:
+        logger.info(f"No variants with {effect_type=}, skipping deduplication")
         return [df]
-    logger.debug("Deduplicating variants (splitting across files)")
-    ea_count: pl.LazyFrame = df.unique(
-        ["ID", "matched_effect_allele"], maintain_order=True
-    ).with_columns([pl.col(["ID"]).cum_count().over(["ID"]).alias("cum_count")])
-
-    # now get groups for each different cumulative count (1 instance, 2, ..., n)
-    groups = ea_count.collect().group_by(["cum_count"], maintain_order=True)
-    group_dfs = (x.lazy() for _, x in groups)
-    ldf_lst = []
 
-    for x in group_dfs:
-        tempdf = df.join(
-            x, on=["row_nr", "ID", "matched_effect_allele"], how="inner"
-        ).select(
-            [
-                "row_nr",
-                "accession",
-                "effect_type",
-                "chr_name",
-                "ID",
-                "matched_effect_allele",
-                "effect_weight",
-            ]
+    logger.debug(
+        f"Splitting {effect_type} variants with different effect alleles and the same ID"
+    )
+
+    # 0: make sure input is sorted for test output to be consistent
+    df = df.sort("row_nr")
+
+    # 1. count the number of unique matched alleles per ID
+    id_alleles_count: pl.DataFrame = (
+        df.unique(["matched_effect_allele", "ID"], maintain_order=True)
+        .with_columns(
+            pl.cum_count("matched_effect_allele").over("ID").alias("cum_count")
         )
+        .collect()
+    )
+
+    # 2: use the cumulative count to create groups to iterate over when splitting
+    groups = id_alleles_count.group_by(["cum_count"], maintain_order=True)
+
+    # 3. unique() will have dropped rows. it's important to keep them
+    dropped_variants: pl.LazyFrame = df.filter(
+        ~pl.col("row_nr").is_in(id_alleles_count.select("row_nr"))
+    )
+
+    ldf_lst = []
+    for i, (_, x) in enumerate(groups):
+        tempdf: pl.LazyFrame = df.filter(
+            pl.col("row_nr").is_in(x.select("row_nr").to_series())
```

Comment:
Why is `row_nr` being used? That refers to the row number within a PGS, and the deduplication here refers to multiple IDs. 

Can we not use the original logic? 

https://github.com/PGScatalog/pgscatalog_utils/blob/7310339af4e58332d75493485d55ef355e2ff3de/pgscatalog_utils/match/write.py#L129-L171